### PR TITLE
feat(x/ecocredit/marketplace): implement buyer and seller fees

### PR DESCRIPTION
**Pull request by aaronc**
## Description

Implementation for #2151 

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: proto/regen/ecocredit/marketplace/v1/tx.proto
Comment (aaronc) [inline proto/regen/ecocredit/marketplace/v1/tx.proto:192]: I switched this to a coin so that the ledger formatting is better with on the Ledger with SIGN_MODE_TEXTUAL. Note that it is still not required, although we could decide to require it when there's a non-zero buyer fee for safety on the user's part - although this will require a marketplace app upgrade

File: x/ecocredit/marketplace/keeper/utils.go
Comment (paul121) [inline x/ecocredit/marketplace/keeper/utils.go:218]: This looks good to me, just a question on how the fees/costs are moving around.

As implemented here a successful `MsgBuyDirect` in a TX with quantity 10, cost 10uregen and both a buyer fee 0.1 and seller fee 0.05, will result in two separate `transfer` events in the message :
1. `transfer` for all total fees (15uregen) from the buyer to the fee pool
2. `transfer` for seller payment (95uregen) from the buyer to the seller

This makes sense and the math adds up. But I wonder if it could lead to some confusion because from the resulting tx it looks like the buyer is paying all of the fees. I'm not sure how important it is too keep this detailed of a record on the TX or if it is much less efficient, but it seems like it could be useful for helpting to calculate total fees spent by buyer or seller (maybe for tax purposes?)

An example:
1. `transfer` for only buyer fees (10uregen) from the buyer to the fee pool
2. `transfer` for seller payment subtotal cost (100uregen) from the buyer to the seller
3. `transfer` for only seller fees (5uregen) from the seller to the fee pool

**Review by paul121 — COMMENTED** [review]
Ryan and I reviewed this PR today during the regen ledger office hours he is hosting and came out with an important question regarding this feature: Why is the seller paying the fee at time of purchase?

An alternative could be paying the seller fee when creating the sell order. There are few reasons this could be desirable:
- There will be a scenario when the seller fee % changes after sell orders have been created. Paying the seller fee at the time of purchase means the seller can never guarantee what fee they might be charged in the future. All sell orders could automatically be closed when changing the seller fee % but this would create a bad UX.
- Paying the seller fee when creating a sell order means the seller would be explicitly consenting to the current seller fee when signing their transaction.
- It would be easier to track the exact seller fee amount (see my above comment)
- This would help to prevent spamming the network of unreasonable sell orders. If a seller cancels the sell order, the fee has already been paid/burned, and benefits the network.

File: x/ecocredit/marketplace/keeper/features/msg_buy_direct.feature
Comment (aaronc) [reply x/ecocredit/marketplace/keeper/features/msg_buy_direct.feature:406]: We can emit the burn event here. Don't have a strong preference.

File: x/ecocredit/marketplace/keeper/utils.go
Comment (aaronc) [reply x/ecocredit/marketplace/keeper/utils.go:218]: Personally, I would stick to the side of efficiency here and not introduce unnecessary state changes because of downstream accounting. If accounting is the consideration, I would rather deal with that by adding attributes to events, maybe enhancing `EventBuyDirect`?

**Comment by aaronc** [comment]
> Ryan and I reviewed this PR today during the regen ledger office hours he is hosting and came out with an important question regarding this feature: Why is the seller paying the fee at time of purchase?
> 
> An alternative could be paying the seller fee when creating the sell order. There are few reasons this could be desirable:
> 
> * There will be a scenario when the seller fee % changes after sell orders have been created. Paying the seller fee at the time of purchase means the seller can never guarantee what fee they might be charged in the future. All sell orders could automatically be closed when changing the seller fee % but this would create a bad UX.
> * Paying the seller fee when creating a sell order means the seller would be explicitly consenting to the current seller fee when signing their transaction.
> * It would be easier to track the exact seller fee amount (see my above comment)
> * This would help to prevent spamming the network of unreasonable sell orders. If a seller cancels the sell order, the fee has already been paid/burned, and benefits the network.

I have to say that I strongly disagree with this. Sellers can change their pricing or close orders at any point in time. Making it so that they pay upfront would mean that the fees need to be escrowed. Also, they are paying for the fees in the sell denom which means they need to gather that capital in advance without having already gotten payment. If fees are low this might not be a problem, but if fees are non-trivial this capital requirement could be a barrier to entry. If we want an anti-spam fee, that should be something different IMHO.

**Comment by paul121** [comment]
Yes, agree that needing the capital in advance could be a barrier to entry, and there could be other solutions for anti-spam fee. But we wondered if these things might be beneficial for the network. It would be a more significant change though.

The fact that fees could change after the sell order is created was our main concern. Perhaps it isn't a huge issue, but just wanted to make sure to flag that. It should at least be clear to the seller that the sell order may incur a fee when it is later closed.

**Comment by aaronc** [comment]
> The fact that fees could change after the sell order is created was our main concern. Perhaps it isn't a huge issue, but just wanted to make sure to flag that. It should at least be clear to the seller that the sell order may incur a fee when it is later closed.

There will always be plenty of advanced notice around governance changes if sellers are concerned about fee changes.

File: x/ecocredit/marketplace/keeper/features/msg_buy_direct.feature
Comment (paul121) [reply x/ecocredit/marketplace/keeper/features/msg_buy_direct.feature:406]: It seems like a burn event should be emitted but it just isn't clear to me if that is already happening. Does the bank keeper do that for us? If so, then maybe we don't need to include in tests.

File: x/ecocredit/marketplace/keeper/utils.go
Comment (paul121) [reply x/ecocredit/marketplace/keeper/utils.go:218]: Yeah, it might make sense to add that in `EventBuyDirect` then. I'm just assuming this would be useful for accounting, maybe we can validate that from a business use-case.

File: x/ecocredit/server/server.go
Comment (JeancarloBarrios) [reply x/ecocredit/server/server.go:70]: I believe it should be its own account, it will also make it easier if it evolves into its own module.

File: x/ecocredit/marketplace/keeper/utils.go
Comment (JeancarloBarrios) [reply x/ecocredit/marketplace/keeper/utils.go:218]: I agree for accounting maybe the event makes more sense

**Review by JeancarloBarrios — APPROVED** [review]
I think it looks good I dont have any major comment LGTM

**Comment by JeancarloBarrios** [comment]
> Yes, agree that needing the capital in advance could be a barrier to entry, and there could be other solutions for anti-spam fee. But we wondered if these things might be beneficial for the network. It would be a more significant change though.
> 
> The fact that fees could change after the sell order is created was our main concern. Perhaps it isn't a huge issue, but just wanted to make sure to flag that. It should at least be clear to the seller that the sell order may incur a fee when it is later closed.

"I like to compare it to the fees you see on the Apple App Store or similar platforms. Essentially, the ecosystem determines what fees are reasonable; under that mindset, what @aaronc suggests makes sense to me. However, I understand the concern of sellers having to adjust prices based on the fee. For quick implementation, it's acceptable because, as @aaronc mentioned, it requires a governance proposal, so it's unlikely to change significantly anytime soon. Later on, we can introduce a custom fee module that allows for more nuanced adjustments.

File: proto/regen/ecocredit/marketplace/v1/tx.proto
Comment (aaronc) [reply proto/regen/ecocredit/marketplace/v1/tx.proto:192]: Is it okay to keep this as not required? @clevinson do you have an opinion?

File: x/ecocredit/marketplace/keeper/utils.go
Comment (aaronc) [reply x/ecocredit/marketplace/keeper/utils.go:219]: @paul121 @JeancarloBarrios added seller and buyer fees paid to `EventBuyDirect`, see https://github.com/regen-network/regen-ledger/pull/2154/commits/b1a7ad233d9a64df09464a547fc7b170fb2ec596. Does that address this sufficiently?

File: proto/regen/ecocredit/marketplace/v1/tx.proto
Comment (clevinson) [reply proto/regen/ecocredit/marketplace/v1/tx.proto:192]: I think it'd be good to have this be required. This way buyers are always consenting to what fees they will pay. In the broader crypto context, users are always consenting to max fees with max gas limits and the like, so I think this makes sense as an approach.

File: x/ecocredit/marketplace/keeper/features/msg_buy_direct.feature
Comment (aaronc) [reply x/ecocredit/marketplace/keeper/features/msg_buy_direct.feature:434]: The bank keeper does include burn events, but since `MsgBuyDirect` now also includes the fees paid and we can assume that if the denom is `uregen` that it was a burn, is it okay if we just leave it at that and not add another event?

File: proto/regen/ecocredit/marketplace/v1/tx.proto
Comment (aaronc) [reply proto/regen/ecocredit/marketplace/v1/tx.proto:190]: I made this required.

File: x/ecocredit/server/server.go
Comment (aaronc) [reply x/ecocredit/server/server.go]: @JeancarloBarrios see 3376e761aad7078597f07e5cadc930fc1a88a791 and caa606ce839bc2ce8069499b5e15f0c7bb2ca3f3 where:
* a marketplace fee pool module account is set up,
* and I added a new `MsgGovSendFromFeePool` message so that governance can actually manage the fee pool funds

Let me know if this all looks good

**Comment by aaronc** [comment]
@paul121 @JeancarloBarrios I believe I have addressed all concerns. I have also added a new message `MsgGovSendFromFeePool` so that governance can actually manage the fee pool. This was initially missing. Please take another look when you get a chance. Once this has two approvals and if there are no objections I think I will go ahead and merge this so we can get closer to a deployment.

File: x/ecocredit/marketplace/keeper/features/msg_buy_direct.feature
Comment (glandua) [reply x/ecocredit/marketplace/keeper/features/msg_buy_direct.feature:435]: What are the pros and cons of adding a specific event here? seems like for users, it is useful to have a specific event.

**Comment by glandua** [comment]
> ## Description
> Implementation for #2151
> 
> ### Author Checklist
> _All items are required. Please add a note to the item if the item is not applicable and please add links to any relevant follow up issues._
> 
> I have...
> 
> * [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
> * [ ]  added `!` to the type prefix if API or client breaking change
> * [ ]  targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
> * [ ]  provided a link to the relevant issue or specification
> * [ ]  followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/building-modules)
> * [ ]  included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
> * [ ]  added a changelog entry to `CHANGELOG.md`
> * [ ]  included comments for [documenting Go code](https://blog.golang.org/godoc)
> * [ ]  updated the relevant documentation or specification
> * [ ]  reviewed "Files changed" and left comments if necessary
> * [ ]  confirmed all CI checks have passed
> 
> ### Reviewers Checklist
> _All items are required. Please add a note if the item is not applicable and please add your handle next to the items reviewed if you only reviewed selected items._
> 
> I have...
> 
> * [ ]  confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
> * [ ]  confirmed `!` in the type prefix if API or client breaking change
> * [ ]  confirmed all author checklist items have been addressed
> * [ ]  reviewed state machine logic
> * [ ]  reviewed API design and naming
> * [ ]  reviewed documentation is accurate
> * [ ]  reviewed tests and test coverage
> * [ ]  manually tested (if applicable)

Note that the cosmos module docs have moved here: https://docs.cosmos.network/main/build/building-modules/intro

**Comment by glandua** [comment]
@aaronc you'd asked me to review this, mentioning that non technical folks like myself could engage in this productively.  So far the only productive comment I have made is finding a broken link to cosmos module docs.  Much of the rest of the conversation i find somewhat hard to know if I have relevant insight or not.

File: x/ecocredit/marketplace/keeper/features/msg_buy_direct.feature
Comment (aaronc) [reply x/ecocredit/marketplace/keeper/features/msg_buy_direct.feature:435]: I don't know if there's really a pro to adding it. The con is that it just adds duplicitous events which takes up extra storage. The bank keeper and marketplace module already emit events containing the relevant data, so I think we should leave it as is.

**Comment by aaronc** [comment]
Thanks @glandua. I think the key high-level points to keep in mind here is that:
* buyers could be charged a percentage fee on marketplace transactions on top of what they're paying for credits in the denon they're buying with
* sellers could be charged a percentage fee off of what they're making from marketplace sales in the denon they're buying with
* governance gets to decide those fees
* fees are only deducted from the buyer and seller when the sale is finalized
* REGEN gets burned and other coins get put in a fee pool
* governance has to decide what to do with the fee pool of non-REGEN tokens

File: x/ecocredit/marketplace/keeper/features/msg_buy_direct.feature
Comment (paul121) [reply x/ecocredit/marketplace/keeper/features/msg_buy_direct.feature:435]: Sorry I might have created some confusion here. I just wanted to make sure that a burn event does exist. I agree we do not need a duplicate event. This would all be very clear if the test scenario included checking that the burn event is emitted but no worries if this is non-trivial to add.

**Comment by paul121** [comment]
> I have also added a new message MsgGovSendFromFeePool so that governance can actually manage the fee pool. This was initially missing. Please take another look when you get a chance. 

I reviewed these changes and they make sense to me. My only question is if we have a way for the marketplace account to grant access  (likely authz authorization) for another account to manage these accrued fees more efficiently than via governance proposals. Can that authorization itself be created via a separate governance proposal? I assume this is already possible or could be added in a followup PR but just want to make sure there is a path for this.

**Comment by aaronc** [comment]
> > I have also added a new message MsgGovSendFromFeePool so that governance can actually manage the fee pool. This was initially missing. Please take another look when you get a chance. 
> 
> 
> 
> I reviewed these changes and they make sense to me. My only question is if we have a way for the marketplace account to grant access  (likely authz authorization) for another account to manage these accrued fees more efficiently than via governance proposals. Can that authorization itself be created via a separate governance proposal? I assume this is already possible or could be added in a followup PR but just want to make sure there is a path for this.

Authz should make this possible with no additional ledger dev needed.

File: x/ecocredit/server/server.go
Comment (JeancarloBarrios) [reply x/ecocredit/server/server.go]: yeah It looks good to me

**Comment by JeancarloBarrios** [comment]
> * governance has to decide what to do with the fee pool of non-REGEN tokens

I believe we can do that already. with authz itself

**Comment by JeancarloBarrios** [comment]
I reviewed it again and it LGTM!!!

**Review by paul121 — APPROVED** [review]

**Comment by aaronc** [comment]
Thanks for the reviews @JeancarloBarrios and @paul121. Going to go ahead and merge this now.

**Pull request merged**
